File: medium/1448. Count Good Nodes in Binary Tree/1448. Count Good Nodes in Binary Tree.go
package main

import "math"

/**
 * Definition for a binary tree node.
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func dfs(root *TreeNode, maxNode int) int {
	if root == nil {
		return 0
	}

	isGood := 0
	if maxNode <= root.Val {
		isGood = 1
	}
	maxNode = max(maxNode, root.Val)

	return isGood + dfs(root.Left, maxNode) + dfs(root.Right, maxNode)
}

func goodNodes(root *TreeNode) int {
	return dfs(root, math.MinInt)
}

File: medium/1448. Count Good Nodes in Binary Tree/1448. Count Good Nodes in Binary Tree.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def goodNodes(self, root: TreeNode) -> int:

        def run(node: Optional[TreeNode], max_node: int) -> int:
            if not node:
                return 0

            isGood = 1 if max_node <= node.val else 0
            max_node = node.val

            return isGood + run(node.left, max_node) + run(node.right, max_node)

        return run(root, 0)
